packages/devtools_shared/lib/src/utils/file_utils.dart

```diff
@@ -2,15 +2,78 @@
-/// should reside in one of the following top-level folders: 'lib', 'bin',
-/// 'integration_test', 'test', or 'benchmark'.
-String packageRootFromFileUriString(String fileUriString) {
-  // TODO(kenz): for robustness, consider sending the root library uri to the
-  // server and having the server look for the package folder that contains the
-  // `.dart_tool` directory.
+/// This method first tries to use the Dart Tooling Daemon to walk the
+/// directory structure of [fileUriString] and look for the package root; we
+/// consider a directory to be a package root if it contains the `.dart_tool`
+/// directory.
+///
+/// If we cannot find the package root using the Dart Tooling Daemon, we use the
+/// heuristic that a Dart executable within a Dart package should reside in one
+/// of the following top-level folders:
+///
+/// * lib
+/// * bin
+/// * integration_test
+/// * test
+/// * benchmark
+/// * example
+///
+/// The URI returned will be a file URI String and will NOT have a trailing
+/// slash.
+Future<String> packageRootFromFileUriString(
+  String fileUriString, {
+  DartToolingDaemon? dtd,
+  @visibleForTesting bool throwOnDtdSearchFailed = false,
+}) async {
+  if (dtd != null) {
+    // Use type [Object] so we can store exceptions of all types.
+    Object? exception;
+
+    final parts = List.of(Uri.parse(fileUriString).pathSegments);
+    while (parts.isNotEmpty) {
+      parts.removeLast();
+      // Include the leading slash because `pathSegments` will remove it, and a
+      // leading slash is required for 'Uri.file' to properly set the 'file'
+      // URI scheme.
+      // Include the trailing slash because the IDE workspace root this URI is
+      // a part of may have a trailing slash, and DTD checks that the URI we
+      // pass to `listDirectoryContents` starts with at least one IDE workspace
+      // root.
+      var pathString = p.joinAll(parts);
+      if (!pathString.startsWith('/')) pathString = '/$pathString';
+      if (!pathString.endsWith('/')) pathString = '$pathString/';
+
+      final currentUri = Uri.file(pathString);
+      try {
+        final directoryContents = await dtd.listDirectoryContents(currentUri);
+        final containsDartToolDirectory = (directoryContents.uris ?? const [])
+            .any((uri) => uri.path.endsWith('.dart_tool/'));
+        if (containsDartToolDirectory) {
+          final uriAsString = currentUri.toString();
+          return uriAsString.endsWith('/')
+              ? uriAsString.substring(0, uriAsString.length - 1)
+              : uriAsString;
+        }
+      } catch (e) {
+        // Fail gracefully on exception, and proceed to using heuristic below.
+        exception = e;
+        break;
+      }
+    }
+    if (throwOnDtdSearchFailed) {
+      throw Exception('Expected DTD to detect the package root, but it failed'
+          '${exception != null ? ' with exception: $exception' : '.'}');
+    }
```

Comment:
You probably shouldn't, but if you call this method with `throwOnDtdSearchFailed=true` and `dtd==null` I think it _won't_ throw (because this is inside the `if` condition), but maybe it should because that's using it incorrectly?

Reply:
this is intentional because this parameter is used to verify that we can properly find the package using only the DTD path. We don't want to throw if the regexp search fails, because we will already be able to verify that by the value returned.

Reply:
Oops, I missed the `@visibleForTesting` on that parameter and thought it might be used by other code :)